Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm

```diff
@@ -662,14 +662,30 @@ static void registerLogHook()
     os_log_set_hook(OS_LOG_TYPE_DEFAULT, ^(os_log_type_t type, os_log_message_t msg) {
         if (msg->buffer_sz > 1024)
             return;
-        char* messageString = os_log_copy_message_string(msg);
-        String logString = String::fromUTF8(messageString);
-        free(messageString);
 
+        String logFormat = String::fromUTF8(msg->format);
         String logChannel = String::fromUTF8(msg->subsystem);
         String logCategory = String::fromUTF8(msg->category);
 
-        callOnMainRunLoop([logChannel = logChannel.isolatedCopy(), logCategory = logCategory.isolatedCopy(), logString = logString.isolatedCopy(), type] {
+        Vector<uint8_t> buffer(msg->buffer, msg->buffer_sz);
+        Vector<uint8_t> privdata(msg->privdata, msg->privdata_sz);
+
+        static NeverDestroyed<Ref<WorkQueue>> queue(WorkQueue::create("Log Queue",  WorkQueue::QOS::Background));
+
+        dispatch_async(queue.get()->dispatchQueue(), [logFormat = logFormat.isolatedCopy(), logChannel = logChannel.isolatedCopy(), logCategory = logCategory.isolatedCopy(), type = type, buffer = WTFMove(buffer), privdata = WTFMove(privdata)] {
+            os_log_message_s msg = { 0 };
+
+            CString format = logFormat.utf8();
+            msg.format = format.data();
+            msg.buffer = buffer.data();
+            msg.buffer_sz = buffer.size();
+            msg.privdata = privdata.data();
+            msg.privdata_sz = privdata.size();
+
+            char* messageString = os_log_copy_message_string(&msg);
+            String logString = String::fromUTF8(messageString);
+            free(messageString);
+
             auto* connection = WebProcess::singleton().existingNetworkProcessConnection();
```

Comment:
You shouldn't be calling `WebProcess::singleton().existingNetworkProcessConnection()` off the main thread as this isn't safe. You're going to have to call it on the main thread and capture a RefPtr<IPC::Connection> in the lambda.